### PR TITLE
🌐 Internationalization for CTA Text for Amp Story Shopping Component

### LINKED ISSUE #36698: [amp story shopping] Set CTA text
### Description

The CTA text should display "Shop Now". 
This will need to be a localized string.
Screen Shot 2021-11-30 at 12 56 43 PM

We might be able to use the [shadowRoot](https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot) API for setting the text content.
shadowRoot.querySelector('.i-amphtml-story-page-attachment-label')

[Docs on i18n](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/_locales/README.md).

### PR DESCRIPTION
**Pull request by jshamble**
Fixes #36698

### CODE REVIEW

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:36]: nit:
What does it update about it? 
Please be as descripte and concise as possible.
 
Adding one word to the function name saves reviewers/maintainers a mental cycle. 
updateCtaText

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:44]: You already have the page ID. Why not use that to get a reference to the page instead of index?

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:40]: A null check is not necessary. 
If no elements are found, the HTMLCollection will be empty.
https://developer.mozilla.org/en-US/docs/Web/API/Element/getElementsByTagName.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:37]: Wouldn't this be plural?

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:44]: If we change the dom structure, this will break.
Do not access by index, access by querySelector.

Accessing by querySelector is also easier to read since we name our elements descriptively.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:55]: Move this outside of mutateElement.
Dom maniuplation should be the only thing in mutateElement.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:52]: nit: 
Just call this localizedString. The enum's descriptive name is only 2 lines below. 
If there were more than one localizedString a longer name would be necessary but it's not needed when it's just one.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:95]: This will be called on every page change. 
We only need this update to happen once.
We only put listeners in layout callback.

How could we approach this to update with SHOPPING_STATE?
A suggestion for how to approach this:
- listen to SHOPPING_STATE
- Check if page includes shopping tag included in the data
- Update CTA text if page has data for the tag

**Review by processprocess — CHANGES_REQUESTED** [review]

File: extensions/amp-story/1.0/_locales/en.json
Comment (processprocess) [inline extensions/amp-story/1.0/_locales/en.json:256]: Concise is better.

"Label for the shopping CTA button when one shopping tag is present."

File: src/service/localization/strings.js
Comment (processprocess) [inline src/service/localization/strings.js:116]: Interactive?

File: src/service/localization/strings.js
Comment (processprocess) [inline src/service/localization/strings.js:115]: It's not the product name though.

**Review by processprocess — CHANGES_REQUESTED** [review]

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:42]: It's not an outlink. Outlinks open a new page.
This can be called ctaButton.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:49]: Why not just use `shoppingTag[0].textContent`? Isn't it the same data?

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:95]: A solution involving listening "on the page change" is provided as it is difficult
to listen and update the CTA text after the shopping state AND wait until the CTA fully loads.
Most likely we will be doing other onPageChange ops in the future. Until a better approach can be found, we are going with the onPageChange Listener.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:49]: I tried this and it was buggy. It would load the data ~ half of the time whereas if I use this it always works.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:95]: I have also tested this a bit further myself and can confirm that as of now, listening for config data changes instead of on the page change will result in buggy (unpredicatble) behavior.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js]: Do not access elements via index. Use a querySelector. 
This is fragile. If the dom structure changes this will break. 

Using class selectors makes the codebase easier to use, for instance we can search `i-amphtml-story-page-open-attachment` and see where it's being modified. 

Please ping me if my comments don't make sense, I'm happy to elaborate!

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:62]: This does not need a variable. Just use `shoppingTags.length`. It's readable.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:79]: We have a `localize` helper we can use. 
Please use this, we can then remove the localizationservice code from this PR.
https://github.com/ampproject/amphtml/blob/556e94a3057960a613a9349926a088a31f2874a5/extensions/amp-story/1.0/amp-story-unsupported-browser-layer.js#L22-L24

My bad for not suggesting this at first. It's brand new. I'm updating the codelab now.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:70]: We must always keep performance in mind when working on our codebase. We cannot have timeouts running.
We need to find a way to only update the button once.
There must be another way. 

The cta button sets it's text using the `data-cta-text` attribute. Maybe we can try setting that attribute if the button is not built.

codepointer to where this happens in when building the CTA:
https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story-open-page-attachment.js#L156

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:111]: There is another way to do this with setting the `data-cta-text` attribute and listening to `SHOPPING_DATA`.
Please give it a try.

Performance is a keystone concept to our codebase. We must do our best to reduce CPU cycles.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js]: It was already updated to use queryselector, I updated the code in [yesterday's commit](https://github.com/ampproject/amphtml/pull/37008/commits/8d673714ae1c654d3488fc6e57109fc8566a0298) - please carefully look through the latest changes (not the review comments as they may point to outdated code such as this one) before commenting. Thanks.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:100]: Please be sure to remove the promise.all

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js]: After a ton of testing and debugging, it turns out that we don't even need to do this - the bug was being caused by the listener not calling the page change sometimes when the page is not loaded.

I added a call to this.updateCtaText_(); in the buildCallback and it now seems to work.

I recall the following comment from `amp-story-auto-ads` (where I got the page change listener pattern from)

"Respond to page navigation event. This method is not called for the first
 page that is shown on load."
   
I'm going to write some unit tests to test this out some more, but I think we've reached a point where we have consistently predictable behavior.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js]: There are still index selectors [here](https://github.com/ampproject/amphtml/blob/ca444fd87d244a20700b78828a10e58ef33d64b7/extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js#L86) and [here](https://github.com/ampproject/amphtml/blob/ca444fd87d244a20700b78828a10e58ef33d64b7/extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js#L63).
Please update these.

Please be sure to apply review comments like this to the entirety of the document!

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js]: I see you marked this as resolved but I'm seeing the page change listener in the code.
Did you try the `data-cta-text` approach? I'm happy to brainstorm with you.

Comment (jshamble) [reply extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js]: I have tried the `data-cta-text` while listening to the shopping data and it seems like it is flaky (only works about half of the time) whereas the page listener seems to always work.

I'll update the code with the `data-cta-text` changes  but it seems like the shopping data subscriber is unreliable.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:68]: Please use `shoppingData` here instead of the store service call.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:76]: We should add a comment here like:
"If CTA button isn't built, set the data-cta-text so text is added when it's built."

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:57]: This variable name is a little confusing because it's a reference to an element inside the CTA button, not the whole button.
Screen Shot 2021-11-23 at 9 04 14 AM
I think we we should mention that it's the anchor el. 
It could be something like `ctaButtonAnchorEl`

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:77]: This needs to be in mutateElement. 
Please wrap the mutateElement on line 79 around this whole if block.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:96]: Please remove the `Promise.resolve`.
We need to await the `Services.storyStoreServiceForOrNull` so that the buildCallback will only proceed once that resolves.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:94]: Please remove LN 91-94 and associated variables (this.container_, this.myText_).
They are not necessary.

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-open-page-attachment.js:275]: To summarize what we went over before the long weekend:
1. This needs to get the text before rendering to the dom. This will likely need to return a promise and render in amp-story-page when that resolves.
2. This logic should only run if attachmentEl is an amp-story-shopping-attachment.

**Comment by processprocess** [comment]
As per the conversation with @newmuis, the i18n service does not currently support placeholders.
This will allow the code to be simplified, we no longer need get a product name and we can use one translation key.

Please update this PR to: 
1. Use one i18n string: "Shop Now". 
2. Use "Shop Now" in both one product and multiple product cases.

This logic can be applied [here](https://github.com/ampproject/amphtml/blob/41864561ced55345ec96cade38baf8e60a52eea6/extensions/amp-story/1.0/amp-story-open-page-attachment.js#L243-L247).

File: src/service/localization/strings.js
Comment (processprocess) [inline src/service/localization/strings.js:115]: nit: 'CTA' in the name could help describe its usage.  
This could be changed to `AMP_STORY_SHOPPING_CTA_LABEL`.

**Review by processprocess — APPROVED** [review]
Approving with one nit, pending @coreymasanto's review.

**Review by coreymasanto — APPROVED** [review]

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js
Comment (jshamble) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:89]: Fixed lol - good to know anyway!

File: extensions/amp-story/1.0/amp-story-open-page-attachment.js
Comment (jshamble) [reply extensions/amp-story/1.0/amp-story-open-page-attachment.js:96]: Good call Corey, I tried doing that but prettier (sadly) optimizes them out. Will leave them out for now I suppose.

**Comment by processprocess** [comment]
Looking back at this PR also. We have a visual test but no unit test.
We should do our best to write both unit and visual tests in case one of the tests are down.

**Pull request merged**
